OsmAnd-java/src/main/java/net/osmand/router/MissingMapsCalculationResult.java

```diff
@@ -24,14 +26,23 @@ public class MissingMapsCalculationResult {
 	private List<WorldRegion> regionsToUpdate;
 	private List<WorldRegion> usedRegions;
 
-	public MissingMapsCalculationResult(RoutingContext missingMapsRoutingContext, List<LatLon> missingMapsPoints) {
+	public MissingMapsCalculationResult(RoutingContext missingMapsRoutingContext, List<LatLon> missingMapsPoints,
+	                                    boolean calledByCalcOnline) {
 		this.missingMapsRoutingContext = missingMapsRoutingContext;
 		this.missingMapsPoints = missingMapsPoints;
+		this.onlineResultPoints = calledByCalcOnline;
 	}
 	
-	public void addMissingMaps(String region) {
-		missingMaps.add(region);
-		mapsToDownload.add(region);
+	public void addMissingMapsByRegionsWithDistinctCountries(List<String> regions) {
```

Comment:
Please separate request, it's not part of original issue

Reply:
It's definitely a part of original issue because Pamir Highway (issue's route belongs to this highway) is located along the border of countries. All the points along the border belong to both regions (in this case, Tajikistan and Afghanistan) and our "too simple" algorithm offers only one country with longer name (Afghanistan) evern when the points are located at Tajikistan's side.